src/bin/stratisd.rs

```diff
@@ -72,54 +97,108 @@ fn run() -> StratisResult<()> {
         }
-        None => 0,
-    };
 
-    // Don't timeout if eventable lets us be event-driven
-    let poll_timeout = if engine_fds_end_idx != 0 { -1 } else { 10000 };
+        // We periodically need to timeout as we are not event driven
+        None => (false, 10000, FD_INDEX_ENGINE),
+    };
 
     loop {
-        // Unconditionally call evented() if engine has no eventable.
-        // This looks like a bad idea, but the only engine that has
-        // no eventable is the sim engine, and for that engine, evented()
-        // is essentially a no-op.
-        if engine_fds_end_idx == 0 {
-            engine.borrow_mut().evented()?;
+        // Process any udev block events
+        if fds[FD_INDEX_UDEV].revents != 0 {
+            loop {
+                match udev.receive_event() {
+                    Some(event) => {
+                        if let Some((devnode, device)) = handle_udev_add(&event) {
+                            let pool_uuid = engine.borrow_mut().block_evaluate(devnode, device);
+
+                            // We need to pretend that we aren't using the variable _pool_uuid so
+                            // that we can conditionally compile out the register_pool when dbus
+                            // is not enabled.
+                            if let Some(_pool_uuid) = pool_uuid {
+                                #[cfg(feature="dbus_enabled")]
+                                libstratis::dbus_api::register_pool(&mut dbus_conn,
+                                                                    &Rc::clone(&engine),
+                                                                    &mut dbus_context,
+                                                                    &mut tree,
+                                                                    _pool_uuid,
+                                                                    &base_object_path)?;
+                            }
+                        }
+                    }
+                    None => {
+                        break;
+                    }
+                };
+            }
         }
 
-        // Handle engine fd, if there is one.
-        // If engine has no eventable, there is no engine fd; in that case
-        // the body of the loop is executed zero times.
-        for pfd in fds[..engine_fds_end_idx]
-                .iter_mut()
-                .filter(|pfd| pfd.revents != 0) {
-            pfd.revents = 0;
-            eventable
-                .as_mut()
-                .expect("index < engine_fds_end_idx <=> index == 0 <=> eventable.is_some()")
-                .clear_event()?;
+        // Handle engine events, if the engine is eventable
+        if engine_eventable {
+            if fds[FD_INDEX_ENGINE].revents != 0 {
+                fds[FD_INDEX_ENGINE].revents = 0;
+
+                eventable
+                    .as_mut()
+                    .expect("eventable.is_some() and engine fd.revents != 0")
```

Comment:
Just "eventable.is_some()", I think. We use the string to explain the logical reason why ```expect()``` can never fail.